Dockerfile

```diff
@@ -11,8 +12,18 @@ RUN python3.11 \
     -mpip install \
     -r /home/getml/requirements.txt
 
+ARG TARGETOS
+ARG TARGETARCH
+
+ENV GETML_VERSION 1.4.0
+
 RUN mkdir /home/getml/.getML /home/getml/.getML/logs /home/getml/.getML/projects /home/getml/demo
-RUN curl https://storage.googleapis.com/static.getml.com/download/1.4.0/getml-1.4.0-x64-linux.tar.gz | tar -C /home/getml/.getML -xvzf -
+RUN if [ "${TARGETARCH}" = "amd64" ]; then \
```

Comment:
Does podman support [heredocs](https://www.docker.com/blog/introduction-to-heredocs-in-dockerfiles/)?

Reply:
Yes.
But...

https://github.com/containers/podman/issues/19578
The closed podman issue.
https://github.com/containers/podman/releases/tag/v4.8.0

https://github.com/containers/buildah/pull/5092
The closed buildah issue.
Buildah v1.33.0 and above
https://github.com/containers/buildah/releases/tag/v1.33.0


I am on:
```bash
$ podman version
Client:       Podman Engine
Version:      4.9.3
API Version:  4.9.3
Go Version:   go1.21.6
Built:        Thu Jan  1 01:00:00 1970
OS/Arch:      linux/amd64
```

```bash
$ buildah version
Version:         1.33.5
Go Version:      go1.21.6
Image Spec:      1.1.0-rc.5
Runtime Spec:    1.1.0
CNI Spec:        1.0.0
libcni Version:  
image Version:   5.29.2
Git Commit:      
Built:           Thu Jan  1 01:00:00 1970
OS/Arch:         linux/amd64
BuildPlatform:   linux/amd64
```

But it dies on me:
```bash
$ podman build . 
STEP 1/22: FROM python:3.11.8
STEP 2/22: RUN useradd getml
--> Using cache c3befa0a4ce19f41dcce3e19cb3778000b0f63512f2e0a12e9ebcbd2cc68c2e8
--> c3befa0a4ce1
STEP 3/22: USER getml
--> Using cache 8f241d0431708778b3505be8068a8140b307cb5ff1a369293be45e6b6f55d4f9
--> 8f241d043170
STEP 4/22: WORKDIR /home/getml
--> Using cache d473a72ab870fdc009ebfd55421e8fa50cc5bdb4b1cf0e115212bdb7af1814b6
--> d473a72ab870
STEP 5/22: COPY --chown=getml:getml --chmod=0777 ./requirements.txt /home/getml/requirements.txt
--> Using cache 33e1f14fea351b998b513a545062dd3bb5d8171b26286017c24fbfc4fb28cea7
--> 33e1f14fea35
STEP 6/22: ENV PATH="$PATH:/home/getml/.local/bin"
--> Using cache 1bc2c722e1f9afcd7aa516810c2e17409a5cd809ae8cfc5eaae3e27f9ed56445
--> 1bc2c722e1f9
STEP 7/22: RUN python3.11     -mpip install     -r /home/getml/requirements.txt
--> Using cache 61adb88ac7c37f0be4727815f2afd5253fb2d8e4e3aa2fad0a7371173d3cf64f
--> 61adb88ac7c3
STEP 8/22: ARG TARGETOS
--> Using cache a10e191fa38a371991c33d208f11243f309e712545a5ad43b7b4b409ee1dd4e9
--> a10e191fa38a
STEP 9/22: ARG TARGETARCH
--> Using cache ddf0deb2ca7b97491c79d473262d7ae861ea841c11ebf77c5db54c000377b62a
--> ddf0deb2ca7b
STEP 10/22: ENV GETML_VERSION 1.4.0
--> Using cache 371ccd5da35f6614a2682545134b6759c4d9657d71d12ab7242eb06d3fe741a8
--> 371ccd5da35f
STEP 11/22: RUN mkdir /home/getml/.getML /home/getml/.getML/logs /home/getml/.getML/projects /home/getml/demo
--> Using cache 03622dedf25dd1d338d08575abb93cff644fb1de8722f8e83aeb597e9859732b
--> 03622dedf25d
STEP 12/22: RUN <<EOT
--> Using cache 076209dc2d133ca0942b131f0c71299c07c464c51f0cdf18d658b72bd440c805
--> 076209dc2d13
STEP 13/22: if [ "${TARGETARCH}" = "amd64" ]; then
ERRO[0000] +(UNHANDLED LOGLEVEL) &imagebuilder.Step{Env:[]string{"TARGETOS=linux", "TARGETARCH=amd64", "PATH=/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/getml/.local/bin", "LANG=C.UTF-8", "GPG_KEY=A035C8C19219BA821ECEA86B64E628F8D684696D", "PYTHON_VERSION=3.11.8", "PYTHON_PIP_VERSION=24.0", "PYTHON_SETUPTOOLS_VERSION=65.5.1", "PYTHON_GET_PIP_URL=https://github.com/pypa/get-pip/raw/dbf0c85f76fb6e1ab42aa672ffca6f0a675d9ee4/public/get-pip.py", "PYTHON_GET_PIP_SHA256=dfe9fd5c28dc98b5ac17979a953ea550cec37ae1b47a5116007395bfacff2ab9", "GETML_VERSION=1.4.0"}, Command:"if", Args:[]string{""}, Flags:[]string{}, Attrs:map[string]bool(nil), Message:"IF ", Original:"if [ \"${TARGETARCH}\" = \"amd64\" ]; then"} 
Error: building at STEP "IF ": Build error: Unknown instruction: "IF" &imagebuilder.Step{Env:[]string{"TARGETOS=linux", "TARGETARCH=amd64", "PATH=/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/getml/.local/bin", "LANG=C.UTF-8", "GPG_KEY=A035C8C19219BA821ECEA86B64E628F8D684696D", "PYTHON_VERSION=3.11.8", "PYTHON_PIP_VERSION=24.0", "PYTHON_SETUPTOOLS_VERSION=65.5.1", "PYTHON_GET_PIP_URL=https://github.com/pypa/get-pip/raw/dbf0c85f76fb6e1ab42aa672ffca6f0a675d9ee4/public/get-pip.py", "PYTHON_GET_PIP_SHA256=dfe9fd5c28dc98b5ac17979a953ea550cec37ae1b47a5116007395bfacff2ab9", "GETML_VERSION=1.4.0"}, Command:"if", Args:[]string{""}, Flags:[]string{}, Attrs:map[string]bool(nil), Message:"IF ", Original:"if [ \"${TARGETARCH}\" = \"amd64\" ]; then"}
```

Inside an extra podman machine it kind of works.
There seems to be a problem with escaping: https://github.com/containers/buildah/issues/5422
Which makes it a little tricky to put export variables as they get either interpreted by podman/buildah as VAR or ENV, or escaped and not removing the escape for the shell part to then have the environment variable extended.
This little hack `\`echo \${GETML_ARCH}\`` seems to work...

Reply:
Cool. I think the `\` hack is fine. But have you tested this on docker as well? :)

Reply:
Argh...

I found the reason, why heredoc doesn't work on Debians podman: 

>   * Revert heredoc and --checksum support in Dockerfiles to workaround ancient docker.io package in Debian, cf #1051202

https://metadata.ftp-master.debian.org/changelogs//main/g/golang-github-containers-buildah/golang-github-containers-buildah_1.33.7+ds1-1_changelog


> Cool. I think the `\` hack is fine. But have you tested this on docker as well? :)

Yes!  
I tested with podman in a `podman machine` and also with docker in [Rancher Desktop](https://rancherdesktop.io/),  which uses also a VM and [moby](https://github.com/moby/moby) inside, which is the open source container engine used inside Docker...
And also with Docker-Desktop, which showed problems with bind-mount. Adjusted that and added hint for other engines.